go/gilded-rose.go

```diff
@@ -1,57 +1,139 @@
-							items[i].quality = items[i].quality + 1
-						}
-					}
-				}
-			}
-		}
+// simple func to return a max of 2 integers
+// useful in this context to stop items value going under 0
+func max(x int, y int) int {
+	if x == y {
+		return x
+	}
+	if x > y {
+		return x
+	}
+	return y
+}
 
-		if items[i].name != "Sulfuras, Hand of Ragnaros" {
-			items[i].sellIn = items[i].sellIn - 1
-		}
+// UpdateQualityBrie handles updates for aged brie
+func UpdateQualityBrie(item *Item) {
+	// always decrement sellInn
+	defer func() {
+		item.sellIn--
+	}()
+
+	// quality cannot go above 50
+	if item.quality == 50 {
+		return
+	}
+
+	// increment when prior to sellIn
+	if item.sellIn > 0 {
+		item.quality++
+	}
+	// increment by 2 if past sellInn date
+	if item.sellIn <= 0 {
+		item.quality = min(item.quality+2, 50)
+	}
+
+}
+
+// UpdateQualityBackstagePasses handles updates for backstage passes
+func UpdateQualityBackstagePasses(item *Item) {
+	// always decrement sellInn
+	defer func() {
+		item.sellIn--
+	}()
+
+	// if past sellInn date then item is worthless
+	if item.sellIn <= 0 {
+		item.quality = 0
+		return
+	}
+
+	// increment by 3 under with under 5 days to go
+	if item.sellIn <= 5 {
+		item.quality = min(item.quality+3, 50)
+		return
+	}
+
+	// increment by 2 under with under 10 days to go
+	if item.sellIn <= 10 {
+		item.quality = min(item.quality+2, 50)
+		return
+	}
+
+	// increment by 1 under normal conditions
+	item.quality++
+
+}
+
+// UpdateQualityStandard handles updates for standard items
+func UpdateQualityStandard(item *Item) {
+	// always decrement sellInn
+	defer func() {
+		item.sellIn--
+	}()
+
+	// if past sellInn date then decrement by 2
+	if item.sellIn <= 0 {
+		item.quality = max(item.quality-2, 0)
+		return
+	}
+
+	// decrement by 1
+	item.quality = max(item.quality-1, 0)
+
+}
+
+// UpdateQualityConjured handles updates for conjured items
+func UpdateQualityConjured(item *Item) {
+	// always decrement sellInn
+	defer func() {
+		item.sellIn--
+	}()
+
+	// if past sellInn date then decrement by 4
+	if item.sellIn <= 0 {
+		item.quality = max(item.quality-4, 0)
+		return
+	}
+
+	// decrement by 2
+	item.quality = max(item.quality-2, 0)
+
+}
+
+// UpdateQuality handles the update of a slice of items. The individual items are handled
+// by separate functions based on the name of the item.
+func UpdateQuality(items []*Item) {
+	for i := 0; i < len(items); i++ {
 
-		if items[i].sellIn < 0 {
-			if items[i].name != "Aged Brie" {
-				if items[i].name != "Backstage passes to a TAFKAL80ETC concert" {
-					if items[i].quality > 0 {
-						if items[i].name != "Sulfuras, Hand of Ragnaros" {
-							items[i].quality = items[i].quality - 1
-						}
-					}
-				} else {
-					items[i].quality = items[i].quality - items[i].quality
-				}
-			} else {
-				if items[i].quality < 50 {
-					items[i].quality = items[i].quality + 1
-				}
-			}
+		switch name := items[i].name; name {
+		case "Sulfuras, Hand of Ragnaros":
```

Comment:
I would add one line comment here explaining  why there is no quality change for sulfuras. It may be easier for 3rd parties to understand what's going on.

Reply:
I did actually do this but only cos empty cases look weird to me. I didnt push that change though for some reason

Reply:
Exactly :) I would even put `break` in there, just to have a hint what is happening.